apps/api/src/deployment/services/top-up-custodial-deployments/top-up-custodial-deployments.service.ts

```diff
@@ -124,27 +124,50 @@ export class TopUpCustodialDeploymentsService implements DeploymentsRefiller {
     const denom = grant.authorization.spend_limit.denom;
     const deploymentLimit = parseFloat(grant.authorization.spend_limit.amount);
 
-    const feesLimit = await this.retrieveFeesLimit(grant.granter, grant.grantee, denom);
-    const { amount } = await this.balanceHttpService.getBalance(grant.granter, denom);
-    const balance = parseFloat(amount);
+    const feesLimit = await this.retrieveFeesLimit(grant.granter, grant.grantee);
+    const [{ amount: balance }, feesBalance] = await Promise.all([
+      this.balanceHttpService.getBalance(grant.granter, denom),
+      denom !== "uakt" && this.balanceHttpService.getBalance(grant.granter, "uakt")
+    ]);
 
     return {
       denom,
       feesLimit,
       deploymentLimit,
-      balance
+      balance,
+      feesBalance: feesBalance?.amount
     };
   }
 
-  private async retrieveFeesLimit(granter: string, grantee: string, denom: string) {
+  private async retrieveFeesLimit(granter: string, grantee: string) {
     const feesAllowance = await this.allowanceHttpService.getFeeAllowanceForGranterAndGrantee(granter, grantee);
-    const feesSpendLimit = feesAllowance.allowance.spend_limit.find(limit => limit.denom === denom);
+    const feesSpendLimit = feesAllowance.allowance.spend_limit.find(limit => limit.denom === "uakt");
 
     return feesSpendLimit ? parseFloat(feesSpendLimit.amount) : 0;
   }
 
-  private canTopUp(amount: number, balances: Pick<Balances, "balance" | "deploymentLimit" | "feesLimit">) {
-    return balances.deploymentLimit > amount && balances.feesLimit > this.MIN_FEES_AVAILABLE && balances.balance > amount + this.MIN_FEES_AVAILABLE;
+  private async calculateTopUpAmount(blockRate: number, balances: Balances) {
+    const amount = await this.drainingDeploymentService.calculateTopUpAmount({ blockRate });
+
+    if (balances.denom === "uakt") {
+      const smallestAmount = Math.min(amount, balances.deploymentLimit - this.MIN_FEES_AVAILABLE, balances.balance - this.MIN_FEES_AVAILABLE);
+      return Math.max(smallestAmount, 0);
+    }
+
+    return Math.min(amount, balances.deploymentLimit, balances.balance);
+  }
+
+  private canTopUp(amount: number, balances: Balances) {
+    if (!amount) {
+      return false;
+    }
+
+    const hasSufficientDeploymentLimit = amount <= balances.deploymentLimit;
+    const hasSufficientFeesLimit = balances.feesLimit >= this.MIN_FEES_AVAILABLE;
+    const hasSufficientFeesBalance = typeof balances.feesBalance === "undefined" || balances.feesBalance >= this.MIN_FEES_AVAILABLE;
```

Comment:
Not sure I understand the diff between balance and limit.

Reply:
Limit as a grant